packages/flutter/test/painting/text_painter_test.dart

```diff
@@ -11,12 +11,121 @@ import 'package:flutter_test/flutter_test.dart';
+  double lastOffset = caretOffset(0);
+  for (int i = 1; i <= text.length; i++) {
+    final double offset = caretOffset(i);
+    expect(offset, greaterThanOrEqualTo(lastOffset));
+    lastOffset = offset;
+  }
+  painter.dispose();
+}
+
+/// Check the caret offsets are accurate for the given single line of LTR text.
+///
+/// This lays out the given text as a single line with [TextDirection.ltr]
+/// and checks the following invariants, which should always hold if the text
+/// is made up of LTR characters:
+///  * The caret offsets go monotonically from 0.0 to the width of the text.
+///  * At each character (that is, grapheme cluster) boundary, the caret
+///    offset equals the width that the text up to that point would have
+///    if laid out on its own.
+///
+/// If you have a [TextSpan] instead of a plain [String],
+/// see [caretOffsetsForTextSpan].
+void checkCaretOffsetsLtr(String text) {
+  final List<int> characterBoundaries = <int>[];
+  final CharacterRange range = CharacterRange.at(text, 0);
+  while (true) {
+    characterBoundaries.add(range.current.length);
+    if (range.stringAfterLength <= 0) {
+      break;
+    }
+    range.expandNext();
+  }
+  _checkCaretOffsetsLtrAt(text, characterBoundaries);
+}
+
+/// Check the caret offsets are accurate for the given single line of LTR text,
+/// ignoring character boundaries within each given cluster.
+///
+/// This concatenates [clusters] into a string and then performs the same
+/// checks as [checkCaretOffsetsLtr], except that instead of checking the
+/// offset-equals-prefix-width invariant at every character boundary,
+/// it does so only at the boundaries between the elements of [clusters].
+///
+/// The elements of [clusters] should be composed of whole characters: each
+/// element should be a valid character range in the concatenated string.
+///
+/// Consider using [checkCaretOffsetsLtr] instead of this function.  If that
+/// doesn't pass, you may have an instance of https://github.com/flutter/flutter/issues/122478 .
+void checkCaretOffsetsLtrFromPieces(List<String> clusters) {
+  final StringBuffer buffer = StringBuffer();
+  final List<int> boundaries = <int>[];
+  boundaries.add(buffer.length);
+  for (final String cluster in clusters) {
+    buffer.write(cluster);
+    boundaries.add(buffer.length);
+  }
+  _checkCaretOffsetsLtrAt(buffer.toString(), boundaries);
+}
+
+/// Compute the caret offsets for the given single line of text, a [TextSpan].
+///
+/// This lays out the given text as a single line with the given [textDirection]
+/// and returns a full list of caret offsets, one at each code unit boundary.
+///
+/// This also checks that the offset at the very start or very end, if the text
+/// direction is RTL or LTR respectively, equals the line's width.
+///
+/// If you have a [String] instead of a nontrivial [TextSpan],
+/// consider using [checkCaretOffsetsLtr] instead.
+List<double> caretOffsetsForTextSpan(TextDirection textDirection, TextSpan text) {
```

Comment:
Nit: Should these functions be private or do you expect them to be used elsewhere?

Reply:
Because this file has a `main` function and will never be imported from, the private/public distinction doesn't mean as much as usual. I think the convention I've mostly seen in test files is to leave off the `_` prefix, even for test fixtures that are firmly private in intention.

Searching, it's a bit more mixed than my impression had been, but I think still more without the `_` prefix than with it: `git grep -P '^(?!import\b|library\b|void main\b|//)\w.*' 'test/*_test.dart'`

Given that and because these read a bit better to me without the `_` prefix, I'm inclined to leave it off.

Reply:
Sounds good, I've also seen both in other tests.